policies/k8s-registry-deprecation/sigstore/deprecated-registry.yaml

```diff
@@ -0,0 +1,28 @@
+apiVersion: policy.sigstore.dev/v1beta1
+kind: ClusterImagePolicy
+metadata:
+  name: deprecated-k8s-grc-io-registry
+  annotations:
+    title: Deprecated registry
+    description: Warn of a registry deprecation
+    learnMoreLink: https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/
+spec:
+  mode: enforce # For warnings, use 'mode: warn'
+  images:
+  - glob: "k8s.gcr.io/**"
+  authorities:
+  - name: k8s-deprecated
+    static:
+      action: pass
+  policy:
+    type: rego
+    data: |
+      package sigstore
+      isCompliant[response] {
+        response := {
+          "result" : true,
+          "error" : "",
+          "warning" : "This repo has been deprecated: https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/"
+        }
+      }
+
```

Comment:
@jimmyraywv There are two options to test policies in Sigstore/policy-controller. We have a CLI, named `policy-tester`, that we use for pre-flight checks against images. Otherwise you need a  k8scluster to deploy the policy-controller and label the namespace where you want to enforce policies there.

* Option 1: Download our `tester` from here (e.g. Darwin): https://github.com/sigstore/policy-controller/releases/download/v0.7.0/tester-darwin-amd64. In the following example, I am checking my policy against the k8s pause image running the following command:

```
 (set -o pipefail && \
    ./tester-darwin-amd64 \
        --policy=sigstore/deprecated-registry.yaml \
        --image=k8s.gcr.io/pause:3.9 | jq)
```

* Option 2: To test your pod spec, you need a kubernetes cluster running. Deploy the policy-controller following these [instructions](https://github.com/sigstore/helm-charts/tree/main/charts/policy-controller), then set the sigstore label `kubectl label ns policy.sigstore.dev/include=true policy-test` to the namespace where you want to enforce policies. 
 Finally, create the policy `kubectl create -f sigstore/deprecated-registry.yaml`. 
 To validate the policy, simply try to create your pod spec:
  
  ```
$ kubectl  create -f podspec.yaml

Error from server (BadRequest): error when creating "podspec.yaml": admission webhook "policy.sigstore.dev" denied the request: validation failed: failed policy: deprecated-k8s-grc-io-registry: spec.containers[1].image
k8s.gcr.io/pause@sha256:7031c1b283388d2c2e09b57badb803c05ebed362dc88d84b480cc47f72a21097 warning: This repo has been deprecated: https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/
no matching policies: spec.containers[0].image
registry.k8s.io/pause@sha256:7031c1b283388d2c2e09b57badb803c05ebed362dc88d84b480cc47f72a21097
```

I can find the error `This repo has been deprecated: https://kubernetes.io/blog/2023/02/06/k8s-gcr-io-freeze-announcement/
no matching policies: spec.containers[0].image
registry.k8s.io/pause@sha256:7031c1b283388d2c2e09b57badb803c05ebed362dc88d84b480cc47f72a21097`

Reply:
Let me know if you need any help.

Reply:
Thanks @hectorj2f. I couldn't get it to work either. When I apply the test-pod with a container from registry.k8s.io, it says that validation failed because there were no matching policies: spec.containers[0].image? Do I need to apply another policy to admit containers from registry.k8s.io? When I launch a pod with a container from k8s.gcr.io it is rejected as designed.

Reply:
The webhook is deny all images because "no matching policies"

Reply:
No, you can use multiple globbing patterns in image.glob, as i show below:

```
images:
  - glob: "registry.k8s.io/**"
  - glob: "k8s.gcr.io/**"
```

Reply:
Sorry, i misunderstood you. You want to accept pods using images from `registry.k8s.io`pods. Forget my last comment.

Reply:
I am not in front of my computer but I will share the steps asap.

Reply:
Incidentally, I also checked the config-policy-controller ConfigMap and no-match-policy is set to warn.

Reply:
Generally, we can create another policy to accept other image globbing patterns that are signed. Or we can configure the policy-controller  to accept images that don't match any policy.

Reply:
By default it rejects images that do not match any policy.

Reply:
> Incidentally, I also checked the config-policy-controller ConfigMap and no-match-policy is set to warn.

@jicowan I figured out your problem. The policy-controller configmap is configured with an example (a placeholder) which is not taken into account (see `_example:` ) by the controller. You need to run the following command to set it to warn.

```
 kubectl patch cm config-policy-controller -n cosign-system --type merge -p '{"data":{"no-match-policy":"warn"}}'
 ```
 
 Once I run that command, I can create any other pods:
 
 ```
kubectl run --image=registry.k8s.io/pause:3.9 -n policy-test mytest
Warning: no matching policies: spec.containers[0].image
Warning: registry.k8s.io/pause@sha256:7031c1b283388d2c2e09b57badb803c05ebed362dc88d84b480cc47f72a21097
pod/mytest created
```

With our default setting of blocking any container image not matching our policies, we prioritize security versus a more  soft or usable experience. Anyway this behavior can be easily configured.

I might propose to remove our **placeholder** (_example:)  in the ConfigMap, so there are not confusions.

Reply:
We just released a new version of the helm chart where the placeholder is commented it out. That should avoid any confusion.